azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,577 @@
+        self.__protocol = protocol
+        self.__ga_family = conf.get_autoupdate_gafamily()
+        self.__autoupdate_enabled = conf.get_autoupdate_enabled()
+        self.__gs_id = self.protocol.get_goal_state().incarnation
+        self._last_attempted_update_time = None
+        self._last_attempted_update_version = FlexibleVersion("0.0.0.0")
+        self._pkg_list = None
+
+    @property
+    def protocol(self):
+        return self.__protocol
+
+    @property
+    def ga_family(self):
+        return self.__ga_family
+
+    @property
+    def autoupdate_enabled(self):
+        return self.__autoupdate_enabled
+
+    @property
+    def gs_id(self):
+        return self.__gs_id
+
+    def __is_update_allowed(self, requested_version):
+        """
+        check to see if update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        """
+        now = time.time()
+
+        if self._last_attempted_update_time is not None and self._last_attempted_update_version == requested_version:
+            next_attempt_time = self._last_attempted_update_time + conf.get_autoupdate_frequency()
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+
+        self._last_attempted_update_time = now
+        self._last_attempted_update_version = requested_version
+        # The time limit elapsed for us to allow updates.
+        return True
+
+    def __get_agent_manifest_from_last_gs(self):
+        """
+        Get the manifests from last GS for the given family
+        Returns: first entry of Manifest
+                 Exception if no manifests found in the last GS
+        """
+        family = self.ga_family
+        agent_family = self.protocol.get_goal_state().extensions_goal_state.agent_families
+        manifests = [m for m in agent_family if m.name == family and len(m.uris) > 0]
+        if len(manifests) == 0:
+            raise Exception(
+                u"No manifest links found for agent family: {0} for incarnation: {1}, skipping agent update".format(
+                    self.ga_family, self.gs_id))
+        return manifests[0]
+
+    def __fetch_pkg_list(self, agent_manifest):
+        """
+        Fetch the agent pkg version, and it's uris from manifest
+        """
+        agent_manifest = self.protocol.get_goal_state().fetch_agent_manifest(agent_manifest.name, agent_manifest.uris)
+        self._pkg_list = agent_manifest.pkg_list
+
+    @staticmethod
+    def __get_requested_version(agent_manifest):
+        """
+        Get the requested version from agent manifest
+        Returns: Requested version if supported and available
+                 None if None of the above is True
+        """
+        if conf.get_enable_ga_versioning() and agent_manifest.is_requested_version_specified:
+            if agent_manifest.requested_version is not None:
+                return FlexibleVersion(agent_manifest.requested_version)
+        return None
```

Comment:
In the new model, the requested version should always be there and if it is missing it is an error. This should raise an exception instead of return None.

Also, let's collect telemetry if the requested version is missing (this should probably be done when fetching the goal state)

Reply:
makes sense